src/cache.rs

```diff
@@ -5,129 +5,100 @@
-            true => {
-                let bytes = fs::read(id_path)?;
-                let mut source = Source::new(&bytes);
-
-                Some(PersistedId::decode(&mut source)?)
-            }
-            false => None,
-        };
-
-        let data = persisted_id.map_or(Ok(Map::new()), |id| id.restore())?;
-        Ok(Self { data })
-    }
-
-    /// Sets the cache data directory. This function can be called once and is a
-    /// requirement for being able to instantiate a cache using [`new`].
-    ///
-    /// # Panics
-    /// If called more than once or if unable to remove old cache file.
-    pub(crate) fn set_data_path(data_dir: PathBuf) -> Result<(), Error> {
-        // remove old cache if it's still there
-        let db_path = data_dir.join("cache.db");
-        if db_path.exists()
-            && !db_path.is_dir()
-            && fs::remove_file(&db_path).is_err()
-        {
-            panic!("Failed to remove old cache files. Please remove {} manually and try again.", db_path.display())
+    /// Returns a new cache instance.
+    pub(crate) fn new<T: AsRef<Path>>(path: T) -> Result<Self, Error> {
+        let db: DB;
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        opts.create_missing_column_families(true);
+        // After 10 million bytes, sort the cache file and create new one
+        opts.set_write_buffer_size(10_000_000);
+
+        let list = DB::list_cf(&Options::default(), &path);
+
+        if list.is_err() {
+            db = DB::open(&opts, path)?;
+        } else {
+            db = DB::open_cf(&opts, path, list?)?;
         }
 
-        CACHE_DATA_PATH
-            .set(data_dir)
-            .expect("cache path can only be set once");
-
-        Persistence::with_backend(&backend(), |_| Ok(()))?;
-
-        Ok(())
-    }
-
-    /// Persist the cache to the internal backend.
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
+        let write_batch = WriteBatch::default();
 
-        persistence_id.encode(&mut sink);
-
-        fs::write(id_path, bytes)?;
-
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    // We store a column family named by hex representation of the psk.
+    // We store key as (psk, note) and value as (KeyData)
+    // The Canon of the tuple, we use it so we can have unique keys per psk and
+    // note
     pub(crate) fn insert(
         &mut self,
         psk: PublicSpendKey,
+        sk: &SecretSpendKey,
         height: u64,
-        note: Option<Note>,
+        note: Note,
```

Comment:
I'm not for changing this.

Having the note as Option, allow you to save the last_height even when the notes do not belong to you.
This means that you don't have to fetch notes starting from the last note owned by you, but instead from the last note produced by anyone

Reply:
Use Option<Note> to keep updating last_height